Makefile

```diff
@@ -16,75 +15,141 @@ BRIGADE_GITHUB_GW_EXTERNAL_PORT := 7744
 IMAGES      = brigade-api brigade-controller brigade-cr-gateway brigade-generic-gateway brigade-vacuum brig brigade-worker git-sidecar
 
-
 .PHONY: echo-images
 echo-images:
 	@echo $(IMAGES)
 
+# This is for debugging the makefile
+.PHONY: list-targets
+list-targets:
+	@$(MAKE) -pRrq -f $(lastword $(MAKEFILE_LIST)) : 2>/dev/null | awk -v RS= -F: '/^# File/,/^# Finished Make data base/ {if ($$1 !~ "^[#.]") {print $$1}}' | sort | egrep -v -e '^[^[:alnum:]]' -e '^$@$$' | xargs
+
 GIT_TAG   = $(shell git describe --tags --always)
 VERSION   ?= ${GIT_TAG}
 IMAGE_TAG ?= ${VERSION}
 LDFLAGS   += -X github.com/Azure/brigade/pkg/version.Version=$(VERSION)
 
+# Set to "amd64 arm64 arm32" to build all archs
+ARCHS ?= amd64
 CX_OSES = linux windows darwin
-CX_ARCHS = amd64
-
-# Build native binaries
-.PHONY: build
-build: $(BINS)
-
-.PHONY: $(BINS)
-$(BINS): vendor
-	go build -ldflags '$(LDFLAGS)' -o bin/$@ ./$@/cmd/$@
-
-# Cross-compile for Docker+Linux
-build-docker-bins: $(addsuffix -docker-bin,$(BINS))
-
-%-docker-bin: vendor
-	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$*/rootfs/$* ./$*/cmd/$*
-
-# To use docker-build, you need to have Docker installed and configured. You should also set
-# DOCKER_REGISTRY to your own personal registry if you are not pushing to the official upstream.
-.PHONY: docker-build
-docker-build: build-docker-bins
-docker-build: $(addsuffix -image,$(IMAGES))
-
-%-image:
-	docker build $(DOCKER_BUILD_FLAGS) -t $(DOCKER_REGISTRY)/$*:$(IMAGE_TAG) $*
-
-# You must be logged into DOCKER_REGISTRY before you can push.
-.PHONY: docker-push
-docker-push: $(addsuffix -push,$(IMAGES))
-
-%-push:
-	docker push $(DOCKER_REGISTRY)/$*:$(IMAGE_TAG)
-
-# Cross-compile binaries for our CX targets.
-# Mainly, this is for brig-cross-compile
-%-cross-compile: vendor
-	@for os in $(CX_OSES); do \
-		echo "building $$os"; \
-		for arch in $(CX_ARCHS); do \
-			GOOS=$$os GOARCH=$$arch CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./bin/$*-$$os-$$arch ./$*/cmd/$*; \
-		done; \
-		if [ $$os = 'windows' ]; then \
-			mv ./bin/$*-$$os-$$arch ./bin/$*-$$os-$$arch.exe; \
-		fi; \
-	done
-
-.PHONY: build-release
-build-release: brig-cross-compile
+
+amd64_GOARCH = amd64
+arm64_GOARCH = arm64
+arm32_GOARCH = arm
+
+amd64_QEMU_IMPORT =
+arm64_QEMU_IMPORT = ADD https://github.com/multiarch/qemu-user-static/releases/download/v3.1.0-2/x86_64_qemu-aarch64-static.tar.gz /usr/bin
+arm32_QEMU_IMPORT = ADD https://github.com/multiarch/qemu-user-static/releases/download/v3.1.0-2/x86_64_qemu-arm-static.tar.gz /usr/bin
+
+amd64_DOCKER_ARCH = amd64
+arm64_DOCKER_ARCH = arm64v8
+arm32_DOCKER_ARCH = arm32v6
+
+amd64_DOCKER_ANNOTATIONS = --os linux --arch amd64
+arm64_DOCKER_ANNOTATIONS = --os linux --arch arm64 --variant v8
+arm32_DOCKER_ANNOTATIONS = --os linux --arch arm
+
+.PHONY: build build-images push-images dockerfiles
+
+define GO_CMD_TARGETS
+# Build command
+$1/rootfs/$2/$1: vendor $(shell find $1 -type f -name '*.go')
```

Comment:
I must admit I'm still trying to figure out _where_ the change is needed :) but I'm seeing these (docker binary) image targets being called when I run `make build`.

I think I prefer maintaining the pattern of `make build` only building go binaries (and placing into `/bin`) corresponding to the os+arch of the host machine.

Then, these docker binary build targets (putting the linux+<arch> binaries in <component>/rootfs) would only be run, as needed, by `make build-images` (and/or, if we had the need to run separately, e.g. via `make build-docker-bins`)

Reply:
Fair enough. I'll try drawing a dependency graph with the names. Then we can discuss how it should work after it's more visible. Makes it easier to make naming changes like this.